nexus/reconfigurator/planning/src/blueprint_builder/builder.rs

```diff
@@ -697,6 +699,88 @@ impl<'a> BlueprintBuilder<'a> {
         Ok(EnsureMultiple::Changed { added: to_add, removed: 0 })
     }
 
+    fn sled_add_zone_external_dns(
+        &mut self,
+        sled_id: SledUuid,
+    ) -> Result<Ensure, Error> {
+        let id = self.rng.zone_rng.next();
+        let ExternalNetworkingChoice {
+            external_ip,
+            nic_ip,
+            nic_subnet,
+            nic_mac,
+        } = self.external_networking.for_new_external_dns()?;
+        let nic = NetworkInterface {
+            id: self.rng.network_interface_rng.next(),
+            kind: NetworkInterfaceKind::Service { id: id.into_untyped_uuid() },
+            name: format!("external-dns-{id}").parse().unwrap(),
+            ip: nic_ip,
+            mac: nic_mac,
+            subnet: nic_subnet,
+            vni: Vni::SERVICES_VNI,
+            primary: true,
+            slot: 0,
+            transit_ips: vec![],
+        };
+
+        let underlay_address = self.sled_alloc_ip(sled_id)?;
+        let http_address =
+            SocketAddrV6::new(underlay_address, DNS_HTTP_PORT, 0, 0);
+        let dns_address = OmicronZoneExternalFloatingAddr {
+            id: self.rng.external_ip_rng.next(),
+            addr: SocketAddr::new(external_ip, DNS_PORT),
+        };
+        let pool_name =
+            self.sled_select_zpool(sled_id, ZoneKind::ExternalDns)?;
+        let dataset = OmicronZoneDataset { pool_name };
+        let zone_type =
+            BlueprintZoneType::ExternalDns(blueprint_zone_type::ExternalDns {
+                dataset,
+                http_address,
+                dns_address,
+                nic,
+            });
+        let filesystem_pool =
+            Some(self.sled_select_zpool(sled_id, zone_type.kind())?);
```

Comment:
What's this for?  I would have expected that we'd only need one pool allocated for this zone and I see that internal DNS seems to use the same pool both for creating the Dataset and for this field.

edit: after some more digging this looks like the pool for the zone's root filesystem.  It makes sense that that could be from a different disk, although I wonder if we're better off choosing the same one anyway.

Reply:
Thanks for opening #6587!

I think I was also unclear on this while doing internal DNS, so I'm now questioning whether that re-use of the pool is correct in `sled_add_zone_internal_dns`. I am not very familiar with ZFS, so I'm still fuzzy on the distinction between a pool and a dataset and for what purposes you'd want one vs the other. The `BlueprintZoneType::InternalDns` struct wants a dataset, but I guess I don't understand why (the field itself is undocumented).

Reply:
And for a little more context: the "pool" is the top-level container that's associated with physical disks (just one U.2 in our case).  Datasets are created on top of them -- each is a separate filesystem (or block device, but we don't use those) with separate configuration, etc.  They're hierarchical sort of like directories so that they can inherit configuration and be managed in groups, etc.

Some of the confusion here may come from the fact that the inventory and blueprint structures talk about a dataset, but they specify it with just the pool name.  I think the way to think about this is:

- What the component needs is one or more *datasets*.  The control plane knows how many it needs and for what purposes.
- The control plane wants control over which disk(s) those datasets are created on.  (That's because each disk has its own policy -- we might decide we want to stop using it, maybe because it's failing, or because an operator tells us they want to replace it, or whatever. )
- The Sled Agent determines (at least to some extent) how datasets are organized on the pool.  For example, it might put all the encrypted storage under a "crypt" dataset and allocate the per-zone datasets from that.

So in the blueprint and in the requests made to sled agents, we specify the configuration for a "dataset"...which actually is just the name of the pool from which the dataset should be allocated.

Reply:
Thank for the detailed explanation! Comit 63e8938f8 makes the durable dataset and root filesystem use the same pool.